### PR TITLE
fix: Allow multiple .periodOffset() offsets

### PR DESCRIPTION
**Pull request by jimgrace**
Without this fix, the indicator .periodOffset() function doesn't work when the same data element or data element operand is specified with more than one period offset value. (See [DHIS2-10551](https://jira.dhis2.org/browse/DHIS2-10551).) This is because the code allowed at most one (nonzero) periodOffset value for each basic DimensionalItemObject (DimensionalItemObjects with the same UID sequence).

### How the code functioned before

DefaultExpressionService.getIndicatorDimensionalItemObjects was called to get the item objects in an indicator expression. After parsing the expression to get the item ids (some of which may contain period offsets), the expression service called DefaultDimensionService.getDataDimensionalItemObjects to translate the itemIds into objects. For performance, this translation process involves three steps: (1) break down the object IDs into their atomic IDs (e.g. DataElementOperandId -> DataElementId and CategoryOptionComboId), (2) for efficiency, look up all atomic objects of the same type at once (e.g. all the DataElements together, and all the CategoryOptionCombos together), and (3) reassemble the original objects from the atomic objects. Step 3 is done in DefaultDimensionService.getItemObjectMap.

The difficulty is that DefaultDimensionService.getItemObjectMap was returning a single object for primitive objects such as DataElement, Indicator, or ProgramIndicator (the object returned from idObjectManager) -- even if that object was referenced by several IDs for that object but with different periodOffset values. So if this method found a nonzero periodOffset in the itemId, it put the nonzero value into the object returned. Therefore, at most one nonzero DataElement, etc. was returned for each data element, etc. id.

By contrast, multiple composite objects like DataElementOperand might be returned for the same (DE, COC) pair, because a new composite object was constructed for this pair each time an itemId referenced the same pair. When the itemId reference a nonzero periodOffset, this periodOffset was copied into the return DataElementOperand object.

After the objects were returned from DefaultDimensionService.getItemObjectMap -> .getDataDimensionalItemObjectMap -> getDataDimensionalItemObjects, where that method filtered the objects to make sure that only one object for each sequence of UIDs remained. If there were two objects with the same UID sequence, the one with a nonzero periodOffset was chosen. If there was more than one nonzero periodOffset, only one would be chosen.

The QueryPlanner (with help from PeriodOffsetUtils) would set up the query to include periods needed for the periodOffset as well as a nonzero periodOffset. Thus, if a given object had both a zero periodOffset and a nonzero periodOffset, both periods would be queried. But if there were multiple (nonzero) periodOffsets for the same item, only one of the nonzero periodOffsets would be considered in planning the periods to query.

Finally, DataHandler.getAggregatedDataValueMap called DefaultExpressionService.getIndicatorDimensionalItemObjects again, and assumed again that only one object will be returned for each UID collection, regardless of how many times the same object was specified with different periodOffsets. (The object will have a nonzero periodOffset if the original expression contained both a zero and nonzero periodOffset.) It then called addItemBasedOnPeriodOffset to add each value to the valueMap needed to find the expression value. If the object has a nonzero periodOffset, DataHandler.addItemBasedOnPeriodOffset created a valueMap entry for the object with the periodOffset, and also a valueMap entry for the same object without the periodOffset (trusting that SerializationUtils.clone will clone the object without the periodOffset). So this would work if the same object were specified with and without a periodOffset. (Note that the code to get the value with the periodOffset assumes that the grid value will be a Double, not an Integer; this code will break for Integer DataElements.)

This is why when there is more than one periodOffset for the same DE/DEO, or other object, only one of them will have values returned. (See the screen shots in [DHIS2-10551](https://jira.dhis2.org/browse/DHIS2-10551).)

### How the code was changed

DefaultDimensionService.getItemObjectMap now returns a different object for every different periodOffset value for objects with the same UID sequence. This is done by making a copy of the DataElement, Indicator, or ProgramIndicator, and setting the object periodOffset in each copied object equal to the object ID's periodOffset. The copy is made through a copy constructor added to each of these objects and their superclasses BaseDataDimensionalItemObject and BaseDimensionalItemObject. **_UPDATE 10-Aug-2021:_** The copies are now made through `SerializationUtils.clone` instead of copy constructors, see comments below from May 11 and Aug 10.

The complex filtering code in DefaultDimensionService.getDataDimensionalItemObjects is eliminated. It just passes through all the objects.

As a consequence, the QueryPlanner now considers all the periodOffsets when planning the query.

DataHandler.getAggregatedDataValueMap now considers all the objects (potentially with different periodOffsets) that match a grid value. This code is greatly simplified and returns valueMap entries only for those objects needed. It doesn't have to provide for a possible zero periodOffset object also when it finds a nonzero periodOffset. And the Integer-breaking code is removed.

### Testing

DimensionServiceTest was updated for the DimensionService changes.

A test for multiple periodOffsets (returning Integer values) was added to AnalyticsServiceTest. However an existing problem was discovered with this test. It calls AnalyticsTestUtils.assertResultGrid to see if grid values match expected values. But it only checked to see if all grid rows contain expected results; it did not check to see if any grid rows were even present. In fact the EventAnalyticsServiceTest was returning no rows from analytics, but the assertResultGrid method was saying that the test passed because there were no rows to check.

AnalyticsTestUtils.assertResultGrid was changed to assert that at least one grid row is returned, and the EventAnalyticsServiceTest was marked with @Ignore and a comment that it needs fixing. Also the test name was added to print if assertResultGrid finds an error, to assist the person debugging the tests.

### Future extensions

Other indicator expression enhancements have been requested. For example, default aggregation type override may need different DataElement objects with properties such as their default aggregation type, and possible aggregation type overrides. Other such enhancements can extend DimensionalItemObject and DimensionalItemId in the same way that periodOffset does.

In fact this problem with periodOffset was discovered by reading the code in order to see how [DHIS2-9831](https://jira.dhis2.org/browse/DHIS2-9831) might be implemented (Indicator expression test for period start/end dates). This is also how [DHIS2-10441](https://jira.dhis2.org/browse/DHIS2-10441) was discovered (Indicator .periodOffset() doesn't work for negative offsets).

### CODE REVIEW

**Review by maikelarabori — APPROVED** [review]
Nice clean-up @jimgrace :+1:

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EventAnalyticsServiceTest.java
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EventAnalyticsServiceTest.java:236]: Maybe better named `scenario` to not be confused with the test method name.

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/dimension/DimensionServiceTest.java
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/dimension/DimensionServiceTest.java:624]: Have a look at `org.hisp.dhis.utils.Assertions#assertContainsOnly`

**Review by jbee — COMMENTED** [review]
Nice analysis. Don't know the specifics of this but it appears your solution has the right approach. What I am not so sure of is the copy-constructor way of creating copies. I could not quite follow why copies are needed. I don't think we use the copy-constructor pattern so far so this feels like a something we should decide on, @larshelge WDYT?

Alternatively maybe it is less invasive (while more brittle) to write copy helpers just for this occasion?

Left some minor suggestions.

File: dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/dimension/DimensionServiceTest.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-core/src/test/java/org/hisp/dhis/dimension/DimensionServiceTest.java]: Nice method. Thanks for writing it. I'm using it here now and I'm sure I will find opportunity to use it elsewhere in the future.

**Comment by jimgrace** [comment]
@jbee Thanks for the review. The copies are needed because: The indicator expression (numerator or denominator) may include several references to a primitive type like _DataElement_ with or without a _periodOffset_, e.g.:

`#{eRwOwCpMzyP}.periodOffset(-2) + #{eRwOwCpMzyP}.periodOffset(-1) + #{eRwOwCpMzyP}`

When parsed, this will generate three different _DimensionalItemId_ values, each with the same data element UID and a different periodOffset. To fetch the _DataElement_ objects from the database given the UIDs, there will be a single data element UID `eRwOwCpMzyP` which the _idObjectManager_ will resolve to a single _DataElement_. But what we desire is three _DataElement_ objects with the same properties from the database but different values of the `transient int periodOffset` (in the _DataElement_ superclass _BaseDimensionalItemObject_). This is so the _QueryPlanner_ will know that it needs to get this data from the reporting period, and also from 1 and 2 periods earlier.

I am also very interested to know if the copy-constructor approach is acceptable or if we should do something different. You said it well that a copy helper might be less intrusive but more brittle. I've made that kind of helper in DHIS2 before, and then another developer updates the base object without realizing that a helper somewhere else should be changed, and the code breaks. So my vote is for more intrusive and less brittle. :)

Another approach could be to use _SerializationUtils.clone_ here. That would be fine with me too, if that is preferred. @larshelge would you like to weigh in on this?

**Comment by jbee** [comment]
Thanks @jimgrace , now understand much better why copy/clones are needed. We have to wait for a decision on how to do the copy/cloning.

**Comment by jimgrace** [comment]
@larshelge I think we're waiting for your opinion on copy-constructors, _SerializationUtils.clone_, or other approach.

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java:645]: Is this a standalone fix which should be applied individually and be backported?

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataelement/DataElement.java
Comment (larshelge) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataelement/DataElement.java:177]: Obviously having "copy constructors" is not ideal, as adds "boilerplate" and is prone to future issues when new properties are added to the data model. Are you able to think of another solution which does not require copying of the domain model @jimgrace ?

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dimension/DefaultDimensionService.java
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dimension/DefaultDimensionService.java:369]: I am sure it is, but is it intentional to remove such a large code block?

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/AnalyticsServiceTest.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/AnalyticsServiceTest.java]: Done.

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EventAnalyticsServiceTest.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EventAnalyticsServiceTest.java]: (See second paragraph of "Testing", above.) I have now returned the EventAnalyticsServiceTest to the way it was, which reports success but only because no data is returned and the check in AnalyticsTestUtils.assertResultGrid doesn't care that no data is present--it only checks correctness for data that is present.

There is no data because it appears that none was ever persisted. The test was never fully written. It parses the test event data but never enters it into the database. See https://github.com/dhis2/dhis2-core/blob/ccee1a68b5075a2b5cfe91c8090b1353025b5995/dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EventAnalyticsServiceTest.java#L246-L271 

I spent a bit of time trying to figure out how to persist the test event data, but I couldn't find an easy piece of code to pattern it after. Given enough time and/or guidance I expect I could figure it out, but this is far into a part of the codebase I've never been before. Perhaps it could be done by someone else and/or in a different ticket?

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/predictor/DefaultPredictionService.java:645]: Good eye. As far as I know this is not a bug fix, but a more robust test for equality that is less prone to breaking in the future. This is in the code that sees if the output dataElement is also present in the sampled data, so values can be carried forward (for example, last month's balance + last month's change = this month's balance. This month's predicted balance is copied back into the input to predict next month's balance, and predictions are always made forward through time in Period order.) As far as I know the code works fine now, both in the unit test (where admittedly it will be the same data element object) and also in operational code. I could break this out into a separate PR if you think that's best, but I don't believe it is currently a bug.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataelement/DataElement.java
Comment (jimgrace) [reply dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataelement/DataElement.java:177]: For why copies are needed, please see my Feb 24 comment above to @jbee. Options include:

1. Copy constructors
Cons:
o Boilerplate code to maintain
Pros:
o Fast
o Minimal special-case code for .periodOffset in DefaultDimensionService
o The copy constructors are at least easy to review and are near the object property definitions

2. Case-specific copy helpers
Cons:
o May not be much simpler than the copy constructors
o May break more easily if, for example, a new (or old) DataElement field is newly used in Analytics that is not in the copy helper. Not near the property definitions.
Pros:
o Fast
o Minimal special-case code for .periodOffset in DefaultDimensionService

3. SerializationUtils.clone, used in DefaultDimensionService.getItemObjectMap only when there is a .periodOffset present (or other query modifier in the future like data element aggregation type override).
Cons:
o Some performance hit, but only when .periodOffset is used
o Some more special-case .periodOffset code in DefaultDimensionService
Pros:
o Simple code, no boilerplate!

4. SerializationUtils.clone, used always in DefaultDimensionService.getItemObjectMap in place of the copy constructors
Cons:
o Some performance hit always
Pros:
o Simple code, no boilerplate!
o Minimal special-case code for .periodOffset in DefaultDimensionService

I think option 4 looks very attractive if the performance hit is negligible. I am willing to do some timing testing to determine this, if you like. We could also consider option 3 as a compromise if the performance hit is not negligible but not too bad.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dimension/DefaultDimensionService.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dimension/DefaultDimensionService.java:369]: Yes, you're right it's intentional. By rethinking the approach to allowing any number of .periodOffset values, all this special-case logic that handles 0 or 1 values is not needed! (See "How the code was changed" above: "The complex filtering...") The cost of fixing the function and eliminating all this special-case logic is that we need some kind of object copying as discussed above.

**Comment by jimgrace** [comment]
I've removed the copy constructors, and am now doing a `SerializationUtils.clone` copy in `DefaultDimensionService`, only when a non-zero `periodOffset` is present (option 3 in my May 11 comment above). So whatever performance hit is incurred by the clone, it is only done when using the `periodOffset` function.

Before this, I investigated further option 5 from May 11. This alternative was less desirable since it would make the `valueMap` more complex. We would need multiple database values for each `DimensionalItemObject` as well as multiple `periodOffsets`.

I also started mocking up code for what could be called an option 6: Use `DimensionalItemId` as the primary object for which to gather data values, and reference the `DimensionalItemObject` from within the `DimensionalItemId`. This approach worked fine in the expression service and `DataHandler`, but hit a majorly failed starting with `DataQueryParams` because all of the dimensions use `DimensionalItemObject`, not just the dimensions in question: `DataElement`, `Indicator`, etc. (I'm sure this is no surprise to Lars!)

So I think the serialization clones are the way to go. Hopefully this PR can be approved now.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/PeriodOffsetUtils.java
Comment (jbee) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/PeriodOffsetUtils.java:196]: Seems `replaceIndexIfMissing` can be removed.

**Review by jbee — APPROVED** [review]
LGTM. I think clone is a good solution for now. Just one cleanup left to do. 
Please wait for @larshelge 's feedback.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dimension/DefaultDimensionService.java
Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dimension/DefaultDimensionService.java:370]: Sounds good.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/PeriodOffsetUtils.java
Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/PeriodOffsetUtils.java]: Done, thanks!

**Pull request merged**
